src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerActor.scala

```diff
@@ -135,34 +139,53 @@ private[impl] class InstanceTrackerActor(
+      case InstanceTrackerActor.StateChanged(ack) =>
+        import context.dispatcher
 
-          case InstanceUpdateEffect.Expunge(instance, events) =>
+        val maybeChange: Future[Option[InstanceChange]] = ack.effect match {
+          case effect @ InstanceUpdateEffect.Update(instance, _, _) =>
+            repository.store(instance) // first store to the repository
+              .map(_ => Some(effect))
+              .recoverWith(tryToRecoverRepositoryFailure(effect))
+              .map(trackerEffect => // update task tracked based on the repository update result
+                trackerEffect.flatMap(ef => {
+                  updateApp(ef.instance.runSpecId, ef.instance.instanceId, newInstance = Some(ef.instance))
+                  Some(InstanceUpdated(ef.instance, lastState = ef.oldState.map(_.state), ef.events))
+                }))
+
+          case effect @ InstanceUpdateEffect.Expunge(instance, _) =>
             logger.debug(s"Received expunge for ${instance.instanceId}")
-            updateApp(instance.runSpecId, instance.instanceId, newInstance = None)
-            Some(InstanceDeleted(instance, lastState = None, events))
+            repository.delete(instance.instanceId) // first store to the repository
+              .map(_ => Some(effect))
+              .recoverWith(tryToRecoverRepositoryFailure(effect))
+              .map(trackerEffect => // update task tracked based on the repository update result
+                trackerEffect.flatMap(ef => {
+                  updateApp(ef.instance.runSpecId, ef.instance.instanceId, newInstance = None)
+                  Some(InstanceDeleted(ef.instance, lastState = None, ef.events))
+                }))
 
           case InstanceUpdateEffect.Noop(_) | InstanceUpdateEffect.Failure(_) =>
-            None
+            Future.successful(None)
         }
 
         val originalSender = sender()
 
         import context.dispatcher
-        maybeChange.map { change =>
-          updateStepProcessor.process(change).recover {
-            case NonFatal(cause) =>
-              // since we currently only use ContinueOnErrorSteps, we can simply ignore failures here
-              logger.warn("updateStepProcessor.process failed", cause)
-              Done
-          }
-        }.getOrElse(Future.successful(Done)).foreach { _ =>
-          ack.sendAck()
-          originalSender ! (())
+        maybeChange.flatMap { changeOpt =>
+          changeOpt.map { change =>
+            updateStepProcessor.process(change).recover {
+              case NonFatal(cause) =>
+                // since we currently only use ContinueOnErrorSteps, we can simply ignore failures here
+                logger.warn("updateStepProcessor.process failed", cause)
+                Done
+            }
+          }.getOrElse(Future.successful(Done))
+        }.onComplete {
+          case Success(_) =>
+            ack.sendAck()
+            originalSender ! (())
+          case Failure(e) =>
```

Comment:
Hm, I think we have to actually wait until this future completes. Aas far as I understand the `InstanceOpProcessor` it sequentualized all updates. We could achieve the effect with  the `WorkQueue`.

Reply:
Please, don't drag the `WorkQueue` here. From what I can see we don't have to wait for this future to complete - the original sender will receive the answer at some point.

Reply:
> if I did not miss anything, in case of failed future, sender never got reply in the previous implementation 

how so? all but Fatal exceptions are ignored (see the comment) with a warning and a `Done` is returned.